### PR TITLE
feat(color-picker): add color-picker component

### LINKED ISSUE #132: ### ColorPicker

### PR DESCRIPTION
**Pull request by Hades-li**
add color-picker component

- [ ] Component Migration

Fix [https://github.com/element-plus/element-plus/issues/132](https://github.com/element-plus/element-plus/issues/132)

### CODE REVIEW

File: packages/color-picker/src/components/alpha-slider.vue [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      const el = instance.vnode.el
+      const alpha = props.color.get('alpha')
+
+      if (!el) return 0
+      return Math.round(alpha * (el.offsetWidth - thumb.value.offsetWidth / 2) / 100)
+    }
+
+    function getThumbTop() {
+      if (!props.vertical) return 0
+      const alpha = color.get('_alpha')
```
Comment (bastarder) [inline packages/color-picker/src/components/alpha-slider.vue:71]: is here correctly ?
i saw the `color.ts` code
```
  get(prop: string) {
    return this['_' + prop]
  }
```
this line will become `color.__alpha` ?
emmm... where is `color` define?  maybe `props.color`?

File: packages/color-picker/src/components/predefine.vue [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    //watch
+    watch(currentColor, val => {
+      const color = new Color()
+      color.fromString(val)
+
+      rgbaColors.value.forEach(item => {
+        item.selected = color.compare(item)
+      })
+    })
+    watch(props.colors, newVal => {
```
Comment (bastarder) [inline packages/color-picker/src/components/predefine.vue:45]: can be reactively?
```suggestion
    watch(() => props.colors, newVal => {
```

File: packages/color-picker/src/components/predefine.vue [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      rgbaColors.value = parseColors(newVal, props.color)
+    })
+    watch(props.color, newVal => {
```
Comment (bastarder) [inline packages/color-picker/src/components/predefine.vue:48]: ditto

File: packages/color-picker/src/components/sv-panel.vue [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      required: true,
+    },
+  },
+  setup(props) {
+    // instance
+    const instance = getCurrentInstance()
+    // data
+    const cursorTop = ref<number>(0)
+    const cursorLeft = ref<number>(0)
+    const background = ref<string>('hsl(0, 100%, 50%)')
```
Comment (bastarder) [inline packages/color-picker/src/components/sv-panel.vue:43]: unnecessary type

File: packages/color-picker/src/components/sv-panel.vue [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      update()
+    })
+    return {
+      cursorTop,
+      cursorLeft,
+      background,
+      colorValue,
+    }
+  },
+  /*data() {
```
Comment (bastarder) [inline packages/color-picker/src/components/sv-panel.vue:106]: remove comments

File: packages/color-picker/src/draggable.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+export default function(element, options) {
+  if (isServer) return
+  const moveFn = function(event) {
+    if (options.drag) {
+      options.drag(event)
+    }
+  }
+  const upFn = function(event) {
+    document.removeEventListener('mousemove', moveFn)
```
Comment (bastarder) [inline packages/color-picker/src/draggable.js:12]: use `on,off`

File: packages/color-picker/src/index.vue [hunk 7]
```diff
@@ -0,0 +1,256 @@
+<template>
+  <el-tooltip
+    v-model="showPicker"
+    effect="light"
+    :manual="true"
```
Comment (bastarder) [inline packages/color-picker/src/index.vue:5]:
```suggestion
    manual
```

File: packages/color-picker/src/index.vue [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    :trigger="['click']"
+    :visible-arrow="false"
+  >
+    <template #content>
+      <div
+        v-click-outside="hide"
+        style="width: 300px"
```
Comment (bastarder) [inline packages/color-picker/src/index.vue:12]: ?

File: packages/color-picker/src/index.vue [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+      >
+        <div class="el-color-dropdown__main-wrapper">
+          <hue-slider
+            ref="hue"
+            :color="color"
+            vertical
+            style="float: right;"
```
Comment (bastarder) [inline packages/color-picker/src/index.vue:19]: ?

File: packages/color-picker/src/index.vue [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    HueSlider,
+    AlphaSlider,
+    ElButton,
+    Predefine,
+  },
+  directives: {
+    ClickOutside,
+  },
+  props: {
+    modelValue: String,
```
Comment (bastarder) [inline packages/color-picker/src/index.vue:126]: modelValue will not be reactively any where?
i saw the element-ui code, they watch value's change

File: packages/color-picker/src/test.vue [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<template>
```
Comment (bastarder) [inline packages/color-picker/src/test.vue:1]: ? remove useless component file

**Review by bastarder — COMMENTED** [review]
you should add unit test

File: packages/color-picker/src/test.vue (hunk 11 above)
Comment (Hades-li) [reply packages/color-picker/src/test.vue]: deleted

File: packages/color-picker/src/components/alpha-slider.vue [hunk 12]
```diff
@@ -0,0 +1,144 @@
+<template>
+  <div class="el-color-alpha-slider" :class="{ 'is-vertical': vertical }">
+    <div
+      ref="bar"
+      class="el-color-alpha-slider__bar"
+      :style="{
+        background: background
```
[the remark this answers is not in the document]
Comment (Hades-li) [reply packages/color-picker/src/components/alpha-slider.vue]: done

File: packages/color-picker/src/components/predefine.vue (hunk 2 above)
Comment (Hades-li) [reply packages/color-picker/src/components/predefine.vue]: fixed:
```
watchEffect(() => {
      rgbaColors.value = parseColors(props.colors, props.color)
})
```

File: packages/color-picker/src/components/predefine.vue (hunk 3 above)
Comment (Hades-li) [reply packages/color-picker/src/components/predefine.vue]: ditto

File: packages/color-picker/src/components/sv-panel.vue (hunk 4 above)
Comment (Hades-li) [reply packages/color-picker/src/components/sv-panel.vue]: done

File: packages/color-picker/src/components/sv-panel.vue (hunk 5 above)
Comment (Hades-li) [reply packages/color-picker/src/components/sv-panel.vue]: done

File: packages/color-picker/src/index.vue (hunk 7 above)
Comment (Hades-li) [reply packages/color-picker/src/index.vue]: done

File: packages/color-picker/src/index.vue (hunk 8 above)
Comment (Hades-li) [reply packages/color-picker/src/index.vue]: done

File: packages/color-picker/src/index.vue (hunk 9 above)
Comment (Hades-li) [reply packages/color-picker/src/index.vue]: ditto

File: packages/color-picker/src/index.vue (hunk 10 above)
Comment (Hades-li) [reply packages/color-picker/src/index.vue:128]: done

File: packages/color-picker/src/draggable.js (hunk 6 above)
Comment (Hades-li) [reply packages/color-picker/src/draggable.js:12]: wait...,i will change draggable.js to ts and resolve this code

Comment (Hades-li) [reply packages/color-picker/src/draggable.js:12]: done

**Comment by Hades-li** [comment]
this pr is closed。change to new pr #375

**Pull request closed without merging**